drivers/storage/portworx/portworx.go

```diff
@@ -679,18 +854,123 @@ func SetPortworxDefaults(toUpdate *corev1.StorageCluster, k8sVersion *version.Ve
+			}
+			if nodeSpecCopy.Storage.CacheDevices == nil && toUpdate.Spec.Storage.CacheDevices != nil {
+				cacheDevices := append(make([]string, 0), *toUpdate.Spec.Storage.CacheDevices...)
+				nodeSpecCopy.Storage.CacheDevices = &cacheDevices
+			}
+			if nodeSpecCopy.Storage.ForceUseDisks == nil && toUpdate.Spec.Storage.ForceUseDisks != nil {
+				nodeSpecCopy.Storage.ForceUseDisks = boolPtr(*toUpdate.Spec.Storage.ForceUseDisks)
+			}
+			if nodeSpecCopy.Storage.JournalDevice == nil && toUpdate.Spec.Storage.JournalDevice != nil {
+				nodeSpecCopy.Storage.JournalDevice = stringPtr(*toUpdate.Spec.Storage.JournalDevice)
+			}
+			if nodeSpecCopy.Storage.SystemMdDevice == nil && toUpdate.Spec.Storage.SystemMdDevice != nil {
+				nodeSpecCopy.Storage.SystemMdDevice = stringPtr(*toUpdate.Spec.Storage.SystemMdDevice)
+			}
+			if nodeSpecCopy.Storage.KvdbDevice == nil && toUpdate.Spec.Storage.KvdbDevice != nil {
+				nodeSpecCopy.Storage.KvdbDevice = stringPtr(*toUpdate.Spec.Storage.KvdbDevice)
+			}
+		}
+
+		if toUpdate.Spec.CloudStorage != nil {
+			if nodeSpec.CloudStorage == nil {
+				nodeSpecCopy.CloudStorage = &corev1.CloudStorageNodeSpec{
+					CloudStorageCommon: *(toUpdate.Spec.CloudStorage.CloudStorageCommon.DeepCopy()),
+				}
+			} else {
+				if nodeSpecCopy.CloudStorage.DeviceSpecs == nil &&
+					toUpdate.Spec.CloudStorage.DeviceSpecs != nil {
+					deviceSpecs := append(make([]string, 0), *toUpdate.Spec.CloudStorage.DeviceSpecs...)
+					nodeSpecCopy.CloudStorage.DeviceSpecs = &deviceSpecs
+				}
+				if nodeSpecCopy.CloudStorage.JournalDeviceSpec == nil &&
+					toUpdate.Spec.CloudStorage.JournalDeviceSpec != nil {
+					nodeSpecCopy.CloudStorage.JournalDeviceSpec = stringPtr(*toUpdate.Spec.CloudStorage.JournalDeviceSpec)
+				}
+				if nodeSpecCopy.CloudStorage.SystemMdDeviceSpec == nil &&
+					toUpdate.Spec.CloudStorage.SystemMdDeviceSpec != nil {
+					nodeSpecCopy.CloudStorage.SystemMdDeviceSpec = stringPtr(*toUpdate.Spec.CloudStorage.SystemMdDeviceSpec)
+				}
+				if nodeSpecCopy.CloudStorage.KvdbDeviceSpec == nil &&
+					toUpdate.Spec.CloudStorage.KvdbDeviceSpec != nil {
+					nodeSpecCopy.CloudStorage.KvdbDeviceSpec = stringPtr(*toUpdate.Spec.CloudStorage.KvdbDeviceSpec)
+				}
+				if nodeSpecCopy.CloudStorage.MaxStorageNodesPerZonePerNodeGroup == nil &&
+					toUpdate.Spec.CloudStorage.MaxStorageNodesPerZonePerNodeGroup != nil {
+					maxStorageNodes := *toUpdate.Spec.CloudStorage.MaxStorageNodesPerZonePerNodeGroup
+					nodeSpecCopy.CloudStorage.MaxStorageNodesPerZonePerNodeGroup = &maxStorageNodes
+				}
+			}
+		}
+
+		updatedNodeSpecs = append(updatedNodeSpecs, *nodeSpecCopy)
+	}
+	toUpdate.Spec.Nodes = updatedNodeSpecs
+}
+
+func setPortworxStorageSpecDefaults(toUpdate *corev1.StorageCluster) {
+	if toUpdate.Spec.Kvdb == nil {
+		toUpdate.Spec.Kvdb = &corev1.KvdbSpec{}
+	}
+	if len(toUpdate.Spec.Kvdb.Endpoints) == 0 {
+		toUpdate.Spec.Kvdb.Internal = true
+	}
```

Comment:
Technically this is not part of storage spec. If you want you can extract it out in `setKvdbDefaults()`.